---
output: github_document
editor_options: 
  chunk_output_type: console
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r echo = FALSE}
knitr::opts_chunk$set(
  warning = FALSE, 
  message = FALSE,
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%",
  cache = TRUE
)
```

# *mulea* - An *R* Package for Multi Enrichment Analysis

## Short Description

*mulea* is a comprehensive overrepresentation and enrichment analyser *R* package, which provides extensive analytical means using diverse databases, statistical models, and *p*-value correction procedures. By using *mulea* we can extend our understanding of the results of various high-throughput -- *e.g.* genomics, transcriptomic, *etc.* -- analyses. Aside from [Gene Ontology](http://geneontology.org/) categories or pathways *mulea* can test enrichment in miRNA or transcription factor regulation, genomic locations, and protein domains, depending on the aim of the study. Alongside the package we provide these [databases]() in a standardised format for many different model organisms ranging from *Escherichia coli* to human, all acquired from publicly available data sources. Moreover, when using *mulea* the user can apply other databases that fit their interest with ease. In the package, we implemented the over-representation analysis (ORA, based on the hypergeometric test) and the gene set enrichment analysis (GSEA, based on the Kolmogorov-Smirnov test) approaches, allowing the user to choose between calculating enrichment for an unranked or ranked lists of elements (*e.g.* genes or proteins). Besides the commonly applied multiple testing correction methods, *mulea* also provides a unique permutation based, *empirical false discovery rate correction* (*eFDR*) of the *p*-values that fits better to the actual dataset and database the user applies. To make the exploration and the biological interpretation of the results easier, the user can also choose among various plot-types when applying the *mulea* package.

## Install Development Version and Load the Package

```{r, eval = FALSE}
library(devtools)
install_github("https://github.com/koralgooll/MulEA.git")
library(MulEA)
```

```{r, echo = FALSE}
library(devtools)
load_all()
```

## Short Introduction About How to Use the *mulea* Package

Import the example data sets:

```{r eval=FALSE}
# import the example gmt database
# in this specific case this is ...
# (to import another gmt database from a GMT file please use the read_gmt() function)
data(example_gmt)

# import a data.frame containing the names of all elements that were involved in 
# the experiment, and their scores (in this case log fold changes) from the experiment
data(example_element_names_pvals)
```

```{r echo=FALSE}
example_gmt <- 
```


## Set Based Test

Define an S4 object of class `ora` (stands for Over-Representation Analysis), run set based tests (hypergeometric test with empirical p-value adjustment) and reshape results:

```{r}
ora_model <- ora(
  gmt = example_gmt,
  element_names = selectDf$select, 
  background_element_names = poolDf$background_element_names,
  p_value_adjustment_method = "PT",
  number_of_permutations = 1000
)

ora_results <- run_test(ora_model)

ora_reshaped_results <- reshape_results(
  model = ora_model, 
  model_results = ora_results, 
  p_value_type_colname='adjustedPValueEmpirical'
)
```

View results:

```{r}
View(ora_results)
```

Plot results:

```{r ora-graph, fig.height = 4}
plot_graph(
  reshaped_results=ora_reshaped_results,
  p_value_max_threshold = 1.00,
  p_value_type_colname = "adjustedPValueEmpirical"
)
```

```{r ora-barplot, fig.height = 3}
plot_barplot(
  reshaped_results = ora_reshaped_results,
  p_value_max_threshold=1.00,
  p_value_type_colname = "adjustedPValueEmpirical"
)
```

```{r ora-heatmap, fig.height = 4}
plot_heatmap(
  reshaped_results=ora_reshaped_results,
  p_value_max_threshold=1.00,
  p_value_type_colname = 'adjustedPValueEmpirical'
)
```

## Ranked Based Test

Define an S4 object of class `RankedBasedTest`, run ranked based test (Subramanian method) and reshape results:

```{r}
ranked_model <- gsea(
  gmt = geneSet,
  element_names = selectDf$select,
  element_scores = selectDf$score
)

ranked_results <- run_test(ranked_model)

ranked_reshaped_results <- reshape_results(
  model = ranked_model, 
  model_results = ranked_results, 
  ontology_id_colname='ontologyId'
)
```

View results:

```{r}
View(ranked_results)
```

Plot results:

```{r eval = FALSE}
# TODO FIX ERROR: Wrongly set data column names
plot_graph(
  reshaped_results=ranked_results,
  p_value_max_threshold = 1.00
)

plot_barplot(
  reshaped_results = ranked_results,
  p_value_max_threshold=1.00
)

plot_heatmap(
  reshaped_results=mulea_sub_reshaped_results,
  p_value_max_threshold=1.00
)
```

# Detailed Description

Functional interpretation of the biological data typically involves identifying key genes, molecules, reactions or pathways by finding non-random changes between two or more conditions or phenotype. Performing enrichment analysis on set of molecules selected from differential omics analysis is a method of choice. Among many packages that can be applied for this task, only few could be applied either to multiple species, ontology types or providing an access to multiple statistics.

MulEA is addressing this gap in addition providing improved way to calculate correction for multiple testing that assume partial dependence between ontology terms and in result limits number of correct associations falsely scored as insignificant. Besides the commonly applied tests, MulEA provides a unique permutation based, empirical false discovery rate correction of the p-values to substitute the too conservative Bonferroni and Benjamini-Hochberg procedures.

MulEA allows enrichment analysis using most popular gene and pathway ontologies (GO, KEGG, Reactome). In addition, one can test enrichment in genomic locations and in gene expression, protein domain, miRNA and transcription factors data bases, all created from publicly available resources and presented in standardized manner. Beyond genes or proteins, MulEA even allows working with basically any kind of data types, i.e. small molecules, chromosome region, enhancers, molecular interactions or any other information defined by the user.

Mulea currently supports 25 organisms from bacteria to human. Because, in addition to knowledge-bases provided alongside the package, the user may provide its own ontology files, MulEA can work with any biological species.

To analyse the data MulEA provide multiple types of statistics in one tool, which allows the user to calculate over-representations using the hypergeometric test, and enrichment analyses of ranked input by the Kolmogorov-Smirnov test.

To conclude, MulEA is a comprehensive enrichment software that allows expansive analyses using diverse ontologies, statistical models and p-value correction procedures that can extend our understanding of the results of various high-throughput analyses and therefore expand our knowledge.

An R-package for fast analysis of bioligical data. The package implements three different approaches of this type of analysis. This file include blueprint of package possibilities, to see more ...
